File: backend/services/news_collector.py
"""Google News RSS를 활용한 최근 7일 이슈 수집."""

from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone
from html import unescape
from typing import Any
from urllib.parse import quote_plus

import feedparser
import httpx

KST = timezone(timedelta(hours=9))


def _strip_html(text: str) -> str:
    if not text:
        return ""
    cleaned = re.sub(r"<[^>]+>", " ", text)
    return unescape(re.sub(r"\s+", " ", cleaned)).strip()


def _parse_published(entry: dict[str, Any]) -> datetime | None:
    for key in ("published_parsed", "updated_parsed"):
        parsed = entry.get(key)
        if parsed:
            return datetime(*parsed[:6], tzinfo=timezone.utc)
    return None


def _within_last_days(dt: datetime | None, days: int = 7) -> bool:
    if dt is None:
        return True
    cutoff = datetime.now(timezone.utc) - timedelta(days=days)
    return dt >= cutoff


def _extract_source(entry: dict[str, Any]) -> str:
    source = entry.get("source", {})
    if isinstance(source, dict) and source.get("title"):
        return source["title"]
    link = entry.get("link", "")
    match = re.search(r"https?://(?:www\.)?([^/]+)", link)
    return match.group(1) if match else "출처 미상"


async def collect_issues(keyword: str, days: int = 7, max_items: int = 12) -> list[dict[str, Any]]:
    """키워드 관련 최근 뉴스/이슈를 수집합니다."""
    query = quote_plus(f"{keyword} when:{days}d")
    url = f"https://news.google.com/rss/search?q={query}&hl=ko&gl=KR&ceid=KR:ko"

    async with httpx.AsyncClient(timeout=20.0, follow_redirects=True) as client:
        response = await client.get(
            url,
            headers={"User-Agent": "IssueReportBot/1.0 (+https://localhost)"},
        )
        response.raise_for_status()
        feed = feedparser.parse(response.text)

    items: list[dict[str, Any]] = []
    for entry in feed.entries:
        published = _parse_published(entry)
        if not _within_last_days(published, days):
            continue

        title = _strip_html(entry.get("title", ""))
        summary = _strip_html(entry.get("summary", entry.get("description", "")))
        link = entry.get("link", "")

        if not title:
            continue

        items.append(
            {
                "title": title,
                "summary": summary or title,
                "url": link,
                "source": _extract_source(entry),
                "published_at": published.astimezone(KST).isoformat() if published else None,
                "published_display": (
                    published.astimezone(KST).strftime("%Y-%m-%d %H:%M")
                    if published
                    else "날짜 미상"
                ),
            }
        )

        if len(items) >= max_items:
            break

    return items

File: lib/news-collector.js
/** Google News RSS 수집 */

const KST_OFFSET_MS = 9 * 60 * 60 * 1000;

function stripHtml(text) {
  if (!text) return '';
  return text
    .replace(/<[^>]+>/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function parsePublished(entry) {
  const raw = entry.pubDate || entry.published || entry.updated || '';
  if (!raw) return null;
  const dt = new Date(raw);
  return Number.isNaN(dt.getTime()) ? null : dt;
}

function withinLastDays(dt, days = 7) {
  if (!dt) return true;
  const cutoff = Date.now() - days * 24 * 60 * 60 * 1000;
  return dt.getTime() >= cutoff;
}

function extractSource(entry) {
  if (entry.source && typeof entry.source === 'object' && entry.source['#text']) {
    return entry.source['#text'];
  }
  if (typeof entry.source === 'string') return entry.source;
  const link = entry.link || '';
  const match = link.match(/https?:\/\/(?:www\.)?([^/]+)/);
  return match ? match[1] : '출처 미상';
}

function formatKst(dt) {
  const kst = new Date(dt.getTime() + KST_OFFSET_MS);
  const y = kst.getUTCFullYear();
  const m = String(kst.getUTCMonth() + 1).padStart(2, '0');
  const d = String(kst.getUTCDate()).padStart(2, '0');
  const h = String(kst.getUTCHours()).padStart(2, '0');
  const min = String(kst.getUTCMinutes()).padStart(2, '0');
  return {
    iso: `${y}-${m}-${d}T${h}:${min}:00+09:00`,
    display: `${y}-${m}-${d} ${h}:${min}`,
  };
}

function parseRssItems(xml) {
  const items = [];
  const itemRegex = /<item>([\s\S]*?)<\/item>/gi;
  let match;
  while ((match = itemRegex.exec(xml)) !== null && items.length < 20) {
    const block = match[1];
    const pick = (tag) => {
      const m = block.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, 'i'));
      return m ? m[1].replace(/<!\[CDATA\[([\s\S]*?)\]\]>/g, '$1').trim() : '';
    };
    const sourceRaw = pick('source');
    items.push({
      title: pick('title'),
      summary: pick('description') || pick('content'),
      link: pick('link'),
      pubDate: pick('pubDate'),
      source: sourceRaw ? stripHtml(sourceRaw) : '',
    });
  }
  return items;
}

export async function collectIssues(keyword, days = 7, maxItems = 12) {
  const query = encodeURIComponent(`${keyword} when:${days}d`);
  const url = `https://news.google.com/rss/search?q=${query}&hl=ko&gl=KR&ceid=KR:ko`;

  const response = await fetch(url, {
    headers: { 'User-Agent': 'IssueReportBot/1.0' },
    redirect: 'follow',
  });
  if (!response.ok) {
    throw new Error(`뉴스 RSS 요청 실패 (${response.status})`);
  }

  const xml = await response.text();
  const entries = parseRssItems(xml);
  const items = [];

  for (const entry of entries) {
    const published = parsePublished(entry);
    if (!withinLastDays(published, days)) continue;

    const title = stripHtml(entry.title);
    const summary = stripHtml(entry.summary || entry.title);
    const link = entry.link || '';
    if (!title) continue;

    const kst = published ? formatKst(published) : null;
    items.push({
      title,
      summary: summary || title,
      url: link,
      source: extractSource(entry),
      published_at: kst?.iso || null,
      published_display: kst?.display || '날짜 미상',
    });

    if (items.length >= maxItems) break;
  }

  return items;
}
